### PR TITLE
Implementation of IPOP CMAES

### PR DESCRIPTION
**Pull request by SuvarshaChennareddy**
This pull request is an initial implementation of IPOP CMAES, which is part of my Google Summer of Code (GSoC) project. Tests have been added, but not all of them are passing.

### CODE REVIEW

**Comment by SuvarshaChennareddy** [comment]
@zoq What should we do about the tests? Keep fine tuning the hyperparameters?

**Comment by zoq** [comment]
> @zoq What should we do about the tests? Keep fine tuning the hyperparameters?

I'm looking into it a little bit more, in the sense of, checking if this is not caused by a hidden bug.

**Comment by SuvarshaChennareddy** [comment]
@zoq Using the Rastrigin Function now for testing, should (hopefully) be more stable.

**Comment by zoq** [comment]
Looks good, let's remove the Schwefel function.

File: tests/ipop_cmaes_test.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  FunctionTest<RosenbrockFunction>(ipopcmaes, 0.1, 0.1);
+}
+
+/**
+ * Run IPOP-Active CMA-ES with the full selection policy on Rastrigin function and
+ * make sure the results are acceptable.
+ */
+TEST_CASE("IPOPActiveCMAESchwefelFunctionTest", "[IPOPCMAESTest]")
+{
+
```
Comment (zoq) [inline tests/ipop_cmaes_test.cpp:60]:
```suggestion
```

Remove the extra empty line.

File: include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    // If the number of function evaluation exceeds the threshold, end 
+    // the optimization.
+    if (nfe.NumFunctionEvaluations() > iterateIn.n_elem * 10000) {
+      Callback::EndOptimization(*this, function, iterate, callbacks...);
+      break;
+    }
+
+    // Increase the population size by the population factor for next restart.
+    cmaes.PopulationSize() *= populationFactor;
+
```
Comment (zoq) [inline include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp:84]: Remove the extra empty line.

File: include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+typename MatType::elem_type IPOPCMAES<CMAESType>::Optimize(
+    SeparableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+
+  StoreBestCoordinates<MatType> sbc;
+  CountFunctionEvaluations nfe;
+  for (size_t i = 0; i < maxRestarts; ++i) {
+
```
Comment (zoq) [inline include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp:68]:
```suggestion
  for (size_t i = 0; i < maxRestarts; ++i)
  {
```

Simple style fix and remove the extra line.

File: include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+template<typename CMAESType>
+template<typename SeparableFunctionType,
+         typename MatType,
+         typename... CallbackTypes>
+typename MatType::elem_type IPOPCMAES<CMAESType>::Optimize(
+    SeparableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+
```
Comment (zoq) [inline include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp:64]: Remove the extra empty line.

File: include/ensmallen_bits/cmaes/ipop_cmaes.hpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  //! Modify the maximum number of restarts.
+  size_t& MaxRestarts() { return maxRestarts; }
+
+  //! Get the CMAES object.
+  const CMAESType& CMAES() const { return cmaes; }
+  //! Modify the CMAES object.
+  CMAESType& CMAES() { return cmaes; }
+
+ private:
+
```
Comment (zoq) [inline include/ensmallen_bits/cmaes/ipop_cmaes.hpp:144]:
```suggestion
```

Remove the extra empty line.

File: include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                                      selectionPolicy,
+                                double stepSize,
+                                const double populationFactor,
+                                const size_t maxRestarts) :
+    cmaes(lambda, transformationPolicy, batchSize, 
+         maxIterations, tolerance, selectionPolicy, stepSize),
+    populationFactor(populationFactor),
+    maxRestarts(maxRestarts)
+{ /* Nothing to do. */  }
+
```
Comment (zoq) [inline include/ensmallen_bits/cmaes/ipop_cmaes_impl.hpp:52]:
```suggestion
```
Remove the extra empty line.

**Review by zoq — COMMENTED** [review]
Some minor style comments.

**Comment by SuvarshaChennareddy** [comment]
> Looks good, let's remove the Schwefel function.

Yup, done👍. Fixed the style as well.

**Comment by SuvarshaChennareddy** [comment]
> Looks like the newly added functions are unstable.

Weird...the tests _seemed_ stable on my system.

**Comment by zoq** [comment]
Did you test it with different random seeds? `ensmallen_tests "[IPOPCMAESTest]" -rng-seed 'time'`

**Comment by SuvarshaChennareddy** [comment]
> Did you test it with different random seeds? `ensmallen_tests "[IPOPCMAESTest]" -rng-seed 'time'`

Yes

**Comment by zoq** [comment]
@mlpack-jenkins test this please

**Comment by SuvarshaChennareddy** [comment]
@zoq I'm sorry, but I'm currently occupied with other responsibilities. However, I'll certainly revisit this when I have more availability.

**Comment by SuvarshaChennareddy** [comment]
@zoq 
I ran a batch script with the following:
`
FOR /L %%i IN (1,1,100) DO (
  ensmallen_tests.exe "[IPOPCMAESTest]" --durations yes --rng-seed "time"
)
`
I got 6 fails (out of a 100 tests)
However, this doesn't seem to be reflected in the ensmallen CPU test. I don't really know how else I could test this.

**Comment by SuvarshaChennareddy** [comment]
It's unable to converge because the step size is diverging.

**Comment by SuvarshaChennareddy** [comment]
@mlpack-jenkins test this please

**Comment by SuvarshaChennareddy** [comment]
@mlpack-jenkins test this please

**Comment by SuvarshaChennareddy** [comment]
@mlpack-jenkins test this please

**Comment by SuvarshaChennareddy** [comment]
@mlpack-jenkins test this please

**Comment by SuvarshaChennareddy** [comment]
@zoq @rcurtin (sorry for the mention) I'm not sure what to do. I fixed some more inconsistencies (hopefully all of them) I found in the CMA-ES implementations. However, there is one test of Demon SGD that keeps failing consistently. I have run the CMA-ES tests locally and they seem to be stable on my machine (a LOT more stable than before, probably because most of the major inconsistencies, if not all, have been fixed). Unfortunately, I don't think I can do anything else. I'd like to hear your thoughts.

**Comment by zoq** [comment]
Happy to look into it.

**Comment by zoq** [comment]
> @zoq did you have a moment to check on what @SuvarshaChennareddy mentioned?

yes, will check

**Pull request closed without merging**